fontir/src/ir.rs

```diff
@@ -115,41 +116,122 @@ pub struct GlyphInstance {
-    pub base: String,
+    pub base: GlyphName,
     /// Affine transformation to apply to the referenced glyph.
     pub transform: Affine,
 }
 
+/// Helps convert points-of-type to a bezier path.
+///
+/// Source formats tend to use streams of point-of-type. Curve manipulation is
+/// often easier on bezier path, so provide a mechanism to convert.
+#[derive(Debug)]
+pub struct GlyphPathBuilder<'a> {
+    glyph_name: &'a str,
+    offcurve: Vec<Point>,
+    path: BezPath,
+    last_move_to: Point,
+}
+
+impl<'a> GlyphPathBuilder<'a> {
+    pub fn new(glyph_name: &'a str) -> GlyphPathBuilder {
+        GlyphPathBuilder {
+            glyph_name,
+            offcurve: Vec::new(),
+            path: BezPath::new(),
+            // kurbo *requires* subpaths start with move so this isn't going to get used
+            last_move_to: Point::ZERO,
+        }
+    }
+
+    fn check_num_offcurve(&self, expected: impl Fn(usize) -> bool) -> Result<(), WorkError> {
+        if !expected(self.offcurve.len()) {
+            return Err(WorkError::InvalidSourceGlyph {
+                glyph_name: self.glyph_name.into(),
+                message: format!(
+                    "{} unexpected consecutive offcurve points {:?}",
+                    self.offcurve.len(),
+                    self.offcurve
+                ),
+            });
+        }
+        Ok(())
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.offcurve.is_empty() && self.path.elements().is_empty()
+    }
+
+    pub fn move_to(&mut self, p: impl Into<Point>) -> Result<(), WorkError> {
+        self.check_num_offcurve(|v| v == 0)?;
+        let p = p.into();
+        self.path.move_to(p);
+        self.last_move_to = p;
+        Ok(())
+    }
+
+    pub fn line_to(&mut self, p: impl Into<Point>) -> Result<(), WorkError> {
+        self.check_num_offcurve(|v| v == 0)?;
+        if self.is_empty() {
+            self.move_to(p)?;
+        } else {
+            self.path.line_to(p);
+        }
+        Ok(())
+    }
+
+    pub fn qcurve_to(&mut self, p: impl Into<Point>) -> Result<(), WorkError> {
+        todo!("Support qcurve to {}", p.into());
+    }
+
+    /// Type of curve depends on accumulated off-curves
+    ///
+    /// <https://unifiedfontobject.org/versions/ufo3/glyphs/glif/#point-types>
+    pub fn curve_to(&mut self, p: impl Into<Point>) -> Result<(), WorkError> {
+        if !self.is_empty() {
+            match self.offcurve.len() {
+                0 => self.path.line_to(p),
+                1 => self.path.quad_to(self.offcurve[0], p.into()),
+                2 => self
+                    .path
+                    .curve_to(self.offcurve[0], self.offcurve[1], p.into()),
+                _ => self.check_num_offcurve(|v| v < 3)?,
+            }
+            self.offcurve.clear();
+        } else {
+            self.move_to(p)?;
+        }
+        Ok(())
+    }
+
+    pub fn offcurve(&mut self, p: impl Into<Point>) -> Result<(), WorkError> {
```

Comment:
Is there a condition where we would expect this to error in the future, or is it about matching the signature of other methods?

Reply:
For symmetry.